crates/bevy_ecs/src/bundle.rs

```diff
@@ -75,26 +77,107 @@ use std::{any::TypeId, collections::HashMap};
 /// bundle, in the _exact_ order that [`Bundle::get_components`] is called.
 /// - [`Bundle::from_components`] must call `func` exactly once for each [`ComponentId`] returned by
 ///   [`Bundle::component_ids`].
-pub unsafe trait Bundle: Send + Sync + 'static {
-    /// Gets this [`Bundle`]'s component ids, in the order of this bundle's [`Component`]s
-    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId>;
-
+pub unsafe trait Bundle: ApplicableBundle + Sized {
     /// Calls `func`, which should return data for each component in the bundle, in the order of
     /// this bundle's [`Component`]s
     ///
     /// # Safety
     /// Caller must return data for each component in the bundle, in the order of this bundle's
     /// [`Component`]s
-    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self
-    where
-        Self: Sized;
+    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self;
+
+    /// Gets this [`Bundle`]'s component ids, in the order of this bundle's [`Component`]s
+    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId>;
 
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This will
     /// [`std::mem::forget`] the bundle fields, so callers are responsible for dropping the fields
     /// if that is desirable.
     fn get_components(self, func: impl FnMut(*mut u8));
```

Comment:
Can `get_components_self` be renamed to `get_components` and this method removed?

Reply:
We cannot do so if we want to keep `ApplicableBundle` sealed. Additionally, it would complicate every implementation of `Bundle` massively, as they would have to cart (ha) around the entire `impl ApplicableBundle for Self`.

Reply:
> We cannot do so if we want to keep ApplicableBundle sealed.

I think there is value in having it not be sealed to allow for dynamically defined bundles like for modding or scripting.

Reply:
Perhaps? The unsealing is not the biggest problem though; that would still be the need to implement two traits to implement `Bundle`.

Reply:
My hope with those suggestions was that it should make it possible to merge both traits.

Reply:
We can't merge the traits is because having a `Bundle` which contains a `Box<dyn Bundle>` would be a problem, as it would break our infrastructure around the bundle graph's caching features, etc.

Reply:
Why would it? Directly forwarding to the inner `dyn Bundle` is fine, right? The bundle graph only maps from `TypeId` to `BundleInfo`, not the other way around.

Reply:
Because the `Box<Dyn Bundle>` could be different for different instances of the containing `Bundle`, and hence we'd get some exceedingly bad UB when the `func` of `get_components` is expecting a different type to the one which is actually given.

Reply:
Right